### ISSUE TITLE
feat: Support NgModules hierarchy sans Lazy Loading

### ISSUE THREAD

**Issue by sirajc** [angular/angular#10958]
**I'm submitting a ...**

```
[X] feature request
```
### Problem statement:

An app with NgModule structure `AppModule` -> `LabsFormsModule` -> `TemplateModule`, `ModelFormControlsModule`, `ModelFormBuilderModule` (3 children)

Each child module has defined routes `RouterModule.forChild(ROUTES)`

**Current behavior**

Now if these modules are imported in `LabsFormsModule` then the routes are flattened and they are loaded in `router-outlet` of `AppModule`. Whereas the intented behavior is that they be loaded in `LabsFormsModule`.
Upon discussion on Gitter, @brandonroberts pointed that this is by design because ParentModule and imported modules shares same context.
Also, Only one injector is created for the ParentModule and imports.

To achive composition, we need to lazy load each of the three modules, then the hierarchy will be maintained.
But these 3 modules are small and are not fit to be loaded lazily, but hierarchy of modules is needed.

Another alternative is, we define Routes on `LabsFormsModule` to load them in `router-outlet` of `LabsFormsModule` but then we need to import component from inside child modules in `forms.routes.ts` file and add them to the routes. This breaks the module concepts.
Components inside child modules are not public and should not be accessed outside of Module. also to be addressable outside of module, components needs to be added to `exports`, In short we are forcing internal implementations to be made public just to support Routing defined in Parent module
See: https://github.com/sirajc/angular2-labs/tree/master/src/app/forms 

**Expected/desired behavior**
### Proposal: Support Hierarchy of modules
1. While defining child module routes

```
RouterModule.forChild(ROUTES, asChildren /*boolean*/)
```

But this has conflicts with context and injector
1. Enable NgModules as children

```
{ path: 'forms',
  component: 'FormsComponent',
  childModules: [ TemplateModule, ModelFormControlsModule, ModelFormBuilderModule ]
}
```
1. Child imports in Parent Module

```
@NgModule({
  imports: [CommonModule],
  children: [TemplateModule, ModelFormControlsModule, ModelFormBuilderModule],
  declarations: [FormsComponent]
})
```

Any of the above 3 approach, which seems fit or any other way Angular can provide, that helps creating NgModule hierarchy and routes composition.

**What is the motivation / use case for changing the behavior?**
By enabling NgModule as children, we can compose application with NgModule hierarchy without the need for lazy loading each and every module.
- **Language:** [all | TypeScript X.X | ES6/7 | ES5]

**Comment by gdi2290** [comment]
I guess this is sort of supported pending rc6 via https://github.com/angular/angular/commit/6b26102931c722e4ddc0038059c1c851427f6374 
```es6
{
  path: 'forms',
  component: 'FormsComponent',
  loadChildren: () => FormsModule
}
// module
@NgModule({
  imports: [
    RouterModule.forChild(ROUTES),
    TemplateModule,
    ModelFormControlsModule,
    ModelFormBuilderModule
  ]
})
class FormsModule {}
```

**Comment by sirajc** [comment]
@gdi2290 This will lazily load LabsFormsModule. and the routes of those 3 imports will flatten out. Making those modules child of AppModule instead of LabsFormsModule. This is the problem we want to solve.

**Comment by brandonroberts** [comment]
@sirajc what Patrick is saying is that you can use loadChildren to synchronously load a module once RC6 lands. So instead of importing the modules into the Forms module, you would have your path for each route and a `loadChildren` entry. An example setup:

**forms.routes.ts**
```ts
import { RouterModule, Route } from '@angular/router';
import { FormsComponent } from './forms.component';
import { TemplateComponent } from './template/template.component';

import { TemplateModule } from './template';

const ROUTES: Route[] = [
  { path: '', component: FormsComponent,
    children: [
      { path: 'template', loadChildren: () => TemplateModule }
    ]
  }
];

export const moduleRoutes = RouterModule.forChild(ROUTES);
```

**template.routes.ts**
```ts
import { Route, RouterModule } from '@angular/router';
import { TemplateComponent } from './template.component';

const ROUTES: Route[] = [
  { path: '', component: TemplateComponent }
];

export const moduleRoutes = RouterModule.forChild(ROUTES);
```

**Comment by gdi2290** [comment]
I guess what I was trying to say is that you can fake lazy load by synchronously returning the module rather than doing anything async.

**Comment by sirajc** [comment]
Thanks, that looks interesting and solves the problem statement. I am on angular-builds, will test it out and update

**Comment by SaltyDH** [comment]
The above solution worked well. I think I understand why it works this way, intuitively, I still want to define  it like this:

Eg. valid routes:
```
/assessment
/assessment/labs
/assessment/results

// AssessmentModule
@NgModule({
    imports: [
	RouterModule.forChild([ { path: 'assessment', component: AssessmentComponent } ])

	LabsModule,
	ResultsModule
    ]
})

// LabsModule
@NgModule({
    imports: [
        RouterModule.forChild([ { path: 'labs', component: LabsComponent } ]
    ]
})

// ResultsModule
@NgModule({
    imports: [
        RouterModule.forChild([	{ path: 'results', component: ResultsComponent } ]
    ]
})

```

It feels more portable, I can import the same module elsewhere as well and the child routes 'automagically' appear as valid child routes of the current module. Sure, route collisions possible but that's what testing is for.

**Comment by sirajc** [comment]
@SaltyDH I disagree with your comment. Flattening of routes as done currently when we import a module is correct. We just need another API with which we can point to modules/import them to load them as child of current module.

**Comment by sirajc** [comment]
@gdi2290 @brandonroberts Thanks
The approach you mentioned does works. `loadChildren: () => TemplateModule` does loads TemplateModule synchronously and routing is composed as needed.

**forms.routes.ts**
```
const ROUTES: Route[] = [
  { path: '', component: FormsComponent,
    children: [
      { path: '', pathMatch: 'full' },
      { path: 'template', loadChildren: () => TemplateModule },
      { path: 'model-form-controls', loadChildren: () => ModelControlsModule },
      { path: 'model-form-builder', loadChildren: () => ModelBuilderModule }
    ]
  }
];
export const moduleRoutes = RouterModule.forChild(ROUTES);
```
**template.routes.ts**
```
const ROUTES: Route[] = [
  { path: '', component: TemplateComponent }
];

export const moduleRoutes = RouterModule.forChild(ROUTES);
```

So the support is already available for this feature. Now we just need the sugar.
Instead of `loadChildren: () => ModelBuilderModule` we need some API that intends the actions that we want, something like `{ path: 'template', module: TemplateModule }`

**Comment by gdi2290** [comment]
cc @vsavkin

**Comment by robwormald** [comment]
We discussed this idea this morning, and we agree the current mechanism is somewhat counterintuitive, though there are use cases for the current behavior, so we'll implement something along these lines.

I like the .defineRoutes idea but I'm sure @vsavkin will have something to say about the logical correctness of it. 

Note you could probably also use the component-less route pattern here:

```ts
RouterModule.forChild([
  { path: 'foo', 
      children: [ ... ]
  }
])
```
It depends if you want to have the parent route defined in the module that contains the routes, or if you want the module that does the importing to define the entry route.

**Comment by vsavkin** [comment]
I think we should not introduce any new syntax sugar before the final is out. Since passing a function to `loadChildren` solves the issue, I'd leave it as is for now.

**Comment by sirajc** [comment]
@vsavkin This sugar will really help strengthen the API. paths can be mapped to module or to the component. With so much emphasis on NgModules and its adoption, to create NgModule per feature etc, this story will say clear the path for us. Moving away from componenets to modules.
Just as we used to map a path to component similarly we can map a path to module. Sounds simple and straightforward
Then saying that NgModule supports lazy loading and you can trick the lazy loading function by by returning a module inside a function of loadChildren.
Please reconsider this sugar.

**Comment by kylecordes** [comment]
There is obvious merit to what @vsavkin  and others have said about not further changing the API between now and final. Who can disagree with that, after all the API churn?

But there is a counter argument. You can think of a project like this in generally open development, as cycling around:

1. Team releases a new version, with new features and therefore API changes - or maybe just fixes and no API changes.
2. Community feedback on the features and API.
3. Repeat

In the long run, I think it the best possible point in this cycle to stabilize and lock in the API is after a cycle of community feedback and minor tweaks. While it seems like the worst possible moment, again thinking in the abstract and not necessarily about the features under discussion here, is to lock in the API right after the core team ships changes but before any community feedback can be incorporated.

**Comment by sirajc** [comment]
We moved directives, pipes, providers - everything to NgModules, y limit routers to components, let there be NgModules as well (keeping lazy loading aside)

**Comment by Martin-Wegner** [comment]
I use the nightly builds and there is a huge problem with synchronous routes. My module has components which use navigate functionality (e.g. Component1 has a link to Component2 via `component2Route` and Component2 has a link to Component1 via `component1Route`). If the module is loaded via route `moduleRoute`, then the currently working routes will be `/moduleRoute/component1Route` and `/moduleRoute/component2Route`. But how should the navigate targets should look like inside the module? If I use a absoulte target `/component1Route`, then the `moduleRoute` is not prefixed and therefore the route is broken. If I use a relative target `component1Route`, then the first result will be `/moduleRoute/component1Route`, which works, but after the next click the route will be `/moduleRoute/component1Route/component2Route`, which not work. So a module must know the loading route or the Angular compiler must prefix every navigate target with the related loading route.

**Comment by brandonroberts** [comment]
@Martin-Wegner that is not a bug. In order to use `router.navigate` to navigate relatively, you must provide the `ActivatedRoute`. A section on this will land in the docs soon.

```ts
import {ActivatedRoute} from '@angular/router';

@Component({...})
class RouteComponent {
  constructor(private router: Router, private route: ActivatedRoute) {}

  go() {
    this.router.navigate(['../somewhere'], { relativeTo: this.route });
  }
}
```

**Comment by Martin-Wegner** [comment]
Oh I know that but I missed the '..' trick :) But I use `routerLink` and here I can't use `relativeTo` :(

**Comment by Martin-Wegner** [comment]
Ah `<a [routerLink]="['../comp2']">Comp2</a>` also works :)

**Comment by brandonroberts** [comment]
@Martin-Wegner `routerLink` don't need `relativeTo` because its implicit

**Comment by crtag** [comment]
Might be late for change indeed, but first of all - I don't see default _flattening_ in action when importing modules with routing, ie inspecting activated route clearly indicates parent/child hierarchy in place. Nor flattening needs to be default strategy on modules routing - what's the advantage of that being default? 
I can't or really don't want to lazy load every single bit of routing just to indicate to application of the routing hierarchy - this feels like weird trick. Having them loaded synchronously is yet another shape of the trick where I still would have to make my parent module explicitly aware of the routing inside imported modules, ie I don't want to change my application module to know what routes are loaded via modules imports, those must be added to existing routing scheme using the strategy I want - either flattening or hierarchical manner.

**Comment by PEsteves8** [comment]
Will there be a way to import a module into another module, without merging its routing? I wanted to use a service that is provided in a module that gets imported to app.module, however by doing that its routes get merged. Currently I'm leaving that service outside of the related imported module and providing it separately in app.module, so that I can have synchronous module loading without messing with the routes.

**Comment by adambuczynski** [comment]
I've run into the same issue. Trying to follow the Routing Module pattern as advertised in the guide (https://angular.io/docs/ts/latest/guide/router.html#!#routing-module), yet finding it not possible to define a route in a Routing Module and then load those as children for a parent route.

**Comment by EvgenyKravchenko** [comment]
Hi guys, 

Is this feature with nested routes supposed to work now?  

I have app routes declared as
```
const routes: Routes = [
  { path: '', redirectTo: 'own', pathMatch: 'full' },
  { path: 'own', component: OwnComponent },
  { path: 'nested', loadChildren: () => NestedModule},
];
export const routing = RouterModule.forRoot(routes);
```
Nested routing declared as 
```
const routes: Routes = [
  { path: 'page1', component: NestedPage1Component },
  { path: 'page2', component: NestedPage2Component },
  { path: '', redirectTo: 'page1', pathMatch: 'full' },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class Module1RoutingModule {}
```

I can get `/own`, `/nested/page1`, `/nested/page2`, but when I try to get root / - I'm getting redirect to `/page1`. Why does that happen, as it's declared in child module with `RouterModule.forChild`, how does it redirects not to  `/own` ?

**Comment by Martin-Wegner** [comment]
@EvgenyKravchenko is this related to my issue https://github.com/angular/angular/issues/12648 ?

**Comment by EvgenyKravchenko** [comment]
@Martin-Wegner maybe, I'm not sure.. I have next imports for app module (import routing before submodule): 

```
imports: [
  routing,
  ...,
  Module1Module
]
```
So, as I understood you have a workaround, but I don't know what can I do here with this structure

**Comment by Martin-Wegner** [comment]
@EvgenyKravchenko in your example my workaround does not work :(

**Comment by brandonroberts** [comment]
@EvgenyKravchenko the routes from imported NgModules are merged together with the host NgModule. This is true whether you're using a routing module or not. There is no hierarchy of routes created just by importing one module into another. That's why you're routes are overriding each other.

**Comment by EvgenyKravchenko** [comment]
@brandonroberts Yeah, I've realized that.. I've fixed this problem with using `loadChildren` and removing any imports of this child module. But I thought that routes which attached as `forChild` shouldn't override routes `forRoot`... Then, what's the difference between them.

**Comment by brandonroberts** [comment]
@EvgenyKravchenko the difference is that `forRoot` adds the providers needed for the `Router` service and its dependencies, which you only need once. `forChild` only provides a means to register additional routes and doesn't add any providers.

**Comment by mp3por** [comment]
@brandonroberts 

> @EvgenyKravchenko the routes from imported NgModules are merged together with the host NgModule. This is true whether you're using a routing module or not. There is no hierarchy of routes created just by importing one module into another. That's why you're routes are overriding each other.

This feels very counter-intuitive to me. I believe that route trees should be auto created in the same way as injector trees are created - via module tree construction. Anything else is too much work or/and counter-intuitive. 

I am sure that I do not fully understand Angular2 so I have created a 'play' project in order to help me and others understand how to create route hierarchies and properly handle unmatched routes. I have done things following the latest tutorial on routing and feature modules (https://angular.io/docs/ts/latest/guide/ngmodule.html). Here is the link: https://github.com/mp3por/Angular2FeatureModulesRouting 

Please review it, lets bring it to life and help everyone :)

**Comment by embarq** [comment]
@mp3por Thanks you alot for your example. Really, thank you very much !

**Comment by arnaudoff** [comment]
Has anybody got a workaround for the issue @adamreisnz mentioned? I'm using the Routing Module pattern as well, and loading the module synchronously via `loadChildren` definitely does not work as expected.

**Comment by brandonroberts** [comment]
@arnaudoff can you provide some sample code or a plunker?

**Comment by Chris-Devine** [comment]
@EvgenyKravchenko Same issue here, the route is now a child and lives at root... not sure how to fix this now

**Comment by brandonroberts** [comment]
@Chris-Devine don't mix the routing module pattern with exporting the `RouterModule.forXX`. NgModules are given priority over ModuleWithProviders, which explains why the child routes would be ordered above the routing. An explanation is also included in the routing guide: https://angular.io/docs/ts/latest/guide/router.html#!#routing-module-order

**Comment by brandonroberts** [comment]
@Chris-Devine here is a working example of @EvgenyKravchenko's plunker: https://plnkr.co/edit/6RhcOAnstlCY64fxnQDM?p=preview

**Comment by DzmitryShylovich** [comment]
@brandonroberts is there any explanation of the benefits of `routing module pattern`?

**Comment by brandonroberts** [comment]
@DzmitryShylovich yep: https://angular.io/docs/ts/latest/cookbook/ngmodule-faq.html#!#routing-module

**Comment by gorshkov-leonid** [comment]
I'm still one of those people who think the current behavior is not intuitive. 
Could you look at the problem from the case. I think, there are two cases of import:
1. I as developer want to import(embed) routing module to own scope. Routing paths will be included to current module.
2. I as developer want to import routing module in new scope (behavior same as in lazy loading). Routing paths will be appended to current module.

Can be organized voting that case is more frequent. I need to implement the second case to low coupling of my application. So workaround described by @adamreisnz is not ideal.

**Comment by gorshkov-leonid** [comment]
I read more carefully the comments above. And tried solution with loadChildren as function. It is important feature. But I have not found it in documentation. Where sync loading was described? 

https://angular.io/docs/ts/latest/guide/ngmodule.html ?
https://angular.io/docs/ts/latest/cookbook/ngmodule-faq.html ?
https://angular.io/docs/ts/latest/api/router/index/Routes-type-alias.html ?
https://angular.io/docs/ts/latest/api/router/index/LoadChildren-type-alias.html ?

**Comment by gorshkov-leonid** [comment]
After AOT compilation I got error: `ERROR Error: Uncaught (in promise): Error: Runtime compiler is not loaded`
I found @robwormald's comment https://github.com/angular/angular/issues/13083#issuecomment-263081191. But the secret is not fully disclosed

**Comment by alexanderbaran** [comment]
@gorshkov-leonid I have the same problem when trying to do aot. I get error: `Error: Error encountered resolving symbol values statically.`

**Comment by gorshkov-leonid** [comment]
may me very smart static analyzer can't infer type of returned value.... ok. But if it cant be used in some cases, why it suggested to use?
you recommended split application with routing to modules. But I want true router with

**Comment by gorshkov-leonid** [comment]
may be very smart static analyzer can't infer type of returned value.... ok. But if it cant be used in some cases, why it suggested to use?
you recommended (official docs) split application with routing to modules for low coupling. I want to grain app, but You did not warn me that it works with lazy routed modules only. Else you cannot include rotes as nested pathes in location . Ok as WA I can use pseudo lazy module (loadchildren with function, promize., etc), but It brokes AOT... may be I should to write ModuleFactory...?  I like angular, but I want more clear, more consistent API, closer to real cases. We try to use it in production. But first impression -yes, it powerful, but ugly. It has huge documantation. But we need periodically research and research tickets.   Why I as user should to type in search field `Runtime compiler is not load`? 

About my question, I saw in youtube that in module can be declared entry components. May be specify entry modules? it is module, that calculated in  loadchildren's function

**Comment by MichaelSpranger** [comment]
@brandonroberts @EvgenyKravchenko  as far as I can see, in the example at plunker (https://plnkr.co/edit/6RhcOAnstlCY64fxnQDM?p=preview) the child content (e.g. `Module1Page1Component`) is displayed in the `router-outlet` of the `App `component, not in that defined in `Module1Component`. Is there currently a way to achieve a nested display? I would find this very useful, and it is also part of the feature request.

**Comment by brandonroberts** [comment]
@MichaelSpranger yes, you can create a nested display by using a parent component in your `Module1` routing configuration. Also if you are synchronously loading your module, you don't also include in your parent module's `imports`

https://plnkr.co/edit/edcOOiQEzXtGZhC29NPP?p=preview

**Comment by acheshkov** [comment]
@brandonroberts thanks, but unfortunately suggested approach does not work for AOT case. i'm surprised that so obvious and logical feature have not implemented. maybe there is a reason for that

**Comment by srix55** [comment]
I spent hours trying to figure this out. I am lazy loading the modules even though I don't want to... When I used @brandonroberts & @gdi2290 's solutions of loadChildren: () => MyModule, I got some weird errors when I did an 
`ng build --prod`
`ERROR in Error encountered resolving symbol values statically. Reference to a local (non-exported) symbol 'appRoutes'. Consider exporting the symbol (position 38:7 in the original .ts file), resolving symbol AppModule in <my-app.module.ts file>`

Trying to fix that seems too hacky to me. So, will leave it at that. Hopefully there's an elegant solution to this soon.

**Comment by gdi2290** [comment]
because of how NGC works you need to export the function so 
loadChildren: () => MyModule

would be 

export function getMyModule() {
  return MyModule;
}

loadChildren: getMyModule

**Comment by manatax** [comment]
@vsavkin just wondering... is this (adding a proper feature to route to a module) going to be added?

**Comment by xi4oc4i** [comment]
Hi all, 

I've one question about the "loadChildren", does it work from the browser address bar?

When I click the link(/detail/id1) on the app,  then click the link(/detail/id1/content/cid1), it works perfectly, but if i access the "/detail/id1/content/cid1" through the browser address bar, it goes to the page, "/detail/id1", any ideas? what have I missed on this configuration?

`const routes: Routes = [
  {
    path: '',
    redirectTo: 'workspace',
    pathMatch: 'full'
  },
  {
    path: 'detail/:id',
    loadChildren: './components/detail.module#DetailModule',
    canActivate: [AuthGuard]
  },
];
@NgModule({
  imports: [
    RouterModule.forRoot(routes)
  ],
  exports: [RouterModule]
})
export class AppRoutingModule { }
------------------------------------------
@NgModule({
  imports: [
    CommonModule,
    FormsModule,
    AccordionModule.forRoot(),
    PopoverModule.forRoot(),
    RouterModule.forChild([{
      path: '',
      component: DetailComponent,
      children: [
        { path: 'content/:Id', component: ContentComponent },
        { path: 'data/:Id', component: DataComponent }
      ]
    }])
  ],
  exports: [DetailComponent]
})
export class DetailModule { }
`

**Comment by aimsoftware** [comment]
I have a use case where I have a module with numerous sub modules. I can get the loadChildren: () => Module to work for one module, but not for many. Has anyone ran into this yet?

Here is the basic module structure for my app:

App Module -> Configuration Module -> [ConfigurationSectionModule1, ConfigurationSectionModule2]

The expected routes would be: 

/configuration/section1/sub-components
/configuration/section2/sub-components

**Comment by shairez** [comment]
See my comment here -

https://github.com/angular/angular/issues/10647#issuecomment-360234116

**Comment by kyranjamie** [comment]
With this solution, the child routes are part of the same `NgModule`.  Functionally, this behaves as desired – routes are not lazyloaded.  Conceptually, however, you may want this separation.  The docs describe `NgModule`s as a "unit of functionality".  So it's understandable to want the encapsulation of a module, without wanting it bundled in a separate lazyloaded file.

**Comment by jbmikk** [comment]
It seems weird to me that depending on whether you lazy load routes or not you get different routes. Shouldn't routes be static regardless of your module loading strategy?

On the other hand I think there should be a separate feature to mount child routes from another module (child module or sibling module, I think this shouldn't be relevant, just as it is irrelevant for components what their origin is). 

Routes should be mounted where the calling module requires them, they could be mounted as flat routes or child routes, pretty much the way components are rendered where they are required.

**Comment by kylecordes** [comment]
Echoing @jbmikk , this is also a pain point when introducing Angular to people coming from other tools. This behavior, where the laziness of the route loading is entangled with the URL scheme, is surprising and offputting.

(I suspect it is merely an artifact of schedule pressure during the initial implementation of lazy loading, and might take a while, certainly a major release, to eventually add a more expected, orthogonal lazy versus URL mechanism. And of course the old behavior should be maintained because there is a **lot** of code out here that relies on it.)

**Comment by atscott** [comment]
Closing as duplicate of #18093. The ergonomics of code splitting and lazy loading is one the top priorities and is one the [roadmap](https://angular.io/guide/roadmap#ergonomic-component-level-code-splitting-apis).
